content/posts/2025-01-15-perf-improvements.md

```diff
@@ -0,0 +1,235 @@
+
+One of the consequences is that we would lose information about each traversed interface. We want to still be able to tell: this sample flow went through `genev_sys_6081`, then `8aa5bd1d532fca8@if2`, and finally `ens5`. So we had to add this information as an array in the map values.
+
+Less flows to fetch thanks to smaller keys, and no user-space deduplication: less used CPU.
+
+**Fig. 1 (again): eBPF agent user-space CPU cores usage, averaged per node**
+
+![User-space CPU]({page.image('perf-improvements-1-8/user-space-cpu.png')})
+
+### Other improvements
+
+There have been other improvements identified, more as low-hanging fruits, but still impactful:
+
+- Removing some redundant calls to `bpf_map_update()`
+- Removing `packed` attribute from the eBPF structures
+
+## Side effects
+
+Those changes triggered a refactoring that doesn't come without consequences and tradeoffs.
+
+- Most importantly, **partial flows**: because now having two different maps, one for the main flows, generated from the TC hook, and another for the enrichments, generated from other hooks, there can sometimes be a mismatch between these two. Especially when the enrichment map keys aren't found in the flows map, the result is a generated partial flow, which is a flow that lacks some information. Namely, the TCP flags, the MAC addresses and the bytes and packets counters are missing in partial flows. It doesn't mean these values are entirely lost; you could still be able to find them in an adjacent flow - it's because flows are evicted periodically, and an eviction might occur precisely at the _wrong moment_ (it's a race condition), with only partial data being available at that time. Another cause for partial flows could be sampling, when the agent is configured with a sampling rate: if for some reason the enrichment data is sampled but the corresponding TC hook packet isn't, this would also result in a partial flow.
+
+**Fig. 8: an example of partial flow, with 0 bytes/packets**
+
+![Partial flow]({page.image('perf-improvements-1-8/partial-flow.png')})
+
+- Limitation in **observed interfaces**: because BPF structure size must be predictable, we cannot store in map all the observed interfaces, we need to set a maximum. This is currently six. If a packet is seen on more than six interfaces, we would only show the first six. Today we consider it sufficient, but we might raise the max later if needed. A Prometheus metrics was added to notify for the maximum reached.
```

Comment:
Reword:
Because BPF structure size must be predictable, we cannot store all the observed interfaces in the map. We need to set a maximum, which is currently six.